tests/kernel/mutex/mutex_api/src/test_mutex_apis.c

```diff
@@ -140,18 +236,171 @@ void test_mutex_lock_unlock(void)
+	zassert_true(thread_ret == TC_FAIL,
+		"waiter thread should still block on the locked mutex");
+
+	zassert_true(k_mutex_unlock(&mutex) == 0, "fail to unlock");
+
+	/* Give thread_waiter a chance to get the mutex */
+	k_sleep(K_MSEC(1));
+
+	/**TESTPOINT: waiter thread got the mutex */
+	zassert_true(thread_ret == TC_PASS,
+			"waiter thread can't take the mutex");
+}
+
+/**
+ * @brief Test mutex's priority inheritance mechanism
+ * @details To verify mutex provide priority inheritance to prevent prority
+ * inversion, and there are 3 cases need to run.
+ * The thread T1 hold the mutex first and cases list as below:
+ * - case 1. When prority T2 > T1, priority inheritance happened.
+ * - case 2. When prority T1 > T2, priority inheritance won't happened.
+ * - case 3. When prority T2 > T3 > T1, priority inheritance happened but T2
+ *   wait for timeout and T3 got the mutex.
+ * @ingroup kernel_mutex_tests
+ */
+void test_mutex_priority_inheritance(void)
+{
+	/**TESTPOINT: run test case 1, given priority T1 < T2 */
+	k_mutex_init(&mutex);
+
+	/* we told thread which case runs now */
+	case_type = 1;
+
+	/* spawn a lower priority thread t1 for holding the mutex */
+	k_thread_create(&tdata, tstack, STACK_SIZE,
+		(k_thread_entry_t)tThread_T1_priority_inheritance,
+			&mutex, &tdata, NULL,
+			K_PRIO_PREEMPT(THREAD_LOW_PRIORITY),
+			K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* wait for spawn thread t1 to take action */
+	k_msleep(TIMEOUT);
+
+	/**TESTPOINT: The current thread does not own the mutex.*/
+	zassert_true(k_mutex_unlock(&mutex) == -EPERM,
+			"fail: current thread does not own the mutex");
+
+	/* spawn a higher priority thread t2 for holding the mutex */
+	k_thread_create(&tdata2, tstack2, STACK_SIZE,
+		(k_thread_entry_t)tThread_T2_priority_inheritance,
+			&mutex, &tdata2, NULL,
+			K_PRIO_PREEMPT(THREAD_HIGH_PRIORITY),
+			K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* wait for spawn thread t2 to take action */
+	k_msleep(TIMEOUT+1000);
+
+	/**TESTPOINT: run test case 2, given priority T1 > T2, this means
+	 * priority inheritance won't happen.
+	 */
+	k_mutex_init(&mutex);
+	case_type = 2;
+
+	/* spawn a lower priority thread t1 for holding the mutex */
+	k_thread_create(&tdata, tstack, STACK_SIZE,
+		(k_thread_entry_t)tThread_T1_priority_inheritance,
+			&mutex, &tdata, NULL,
+			K_PRIO_PREEMPT(THREAD_HIGH_PRIORITY),
+			K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* wait for spawn thread t1 to take action */
+	k_msleep(TIMEOUT);
+
+	/* spawn a higher priority thread t2 for holding the mutex */
+	k_thread_create(&tdata2, tstack2, STACK_SIZE,
+		(k_thread_entry_t)tThread_T2_priority_inheritance,
+			&mutex, &tdata2, NULL,
+			K_PRIO_PREEMPT(THREAD_LOW_PRIORITY),
+			K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* wait for spawn thread t2 to take action */
+	k_msleep(TIMEOUT+1000);
+
+	/**TESTPOINT: run test case 3, given priority T1 < T3 < T2, but t2 do
+	 * not get mutex due to timeout.
+	 */
+	k_mutex_init(&mutex);
+	case_type = 3;
```

Comment:
How about creating three separate test functions? This function is very long

Reply:
@sjg20 ok, I will split it to three test functions in the next submit, thanks!